File: test/unit/features/gigs/gig_search_index_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mube/src/features/gigs/data/gig_search_index.dart';

void main() {
  group('gig search index', () {
    test('normalizes accents, punctuation and repeated whitespace', () {
      expect(
        normalizeGigSearchText('  São José / MÚSICA!  '),
        'sao jose musica',
      );
    });

    test('includes title, description and location label', () {
      expect(
        buildGigSearchText(
          title: 'Banda de baile',
          description: 'Show corporativo',
          location: const {'label': 'Niterói, RJ'},
        ),
        'banda de baile show corporativo niteroi rj',
      );
    });

    test('builds stable one, two and three character grams', () {
      final grams = buildGigSearchGrams(title: 'Sax', description: 'Jazz');

      expect(grams, containsAll(<String>['s', 'sa', 'sax', 'j', 'ja', 'jaz']));
      expect(grams.toSet().length, grams.length);
      expect(grams, orderedEquals([...grams]..sort()));
    });

    test('uses the first normalized trigram as the indexed lookup key', () {
      expect(gigSearchLookupGram('  São Paulo '), 'sao');
      expect(gigSearchLookupGram('DJ'), 'dj');
      expect(gigSearchLookupGram('!!!'), isEmpty);
    });

    test('caps indexed text to keep Firestore documents bounded', () {
      final text = buildGigSearchText(
        title: List.filled(3000, 'a').join(),
        description: 'description',
      );

      expect(text, hasLength(2500));
    });
  });
}

File: functions/test/gig_search_index.test.ts
import {
  buildGigSearchGrams,
  GIG_SEARCH_SCHEMA_VERSION,
  normalizeGigSearchText,
} from "../src/gig_search_index";

describe("gig search index", () => {
  test("normalizes Brazilian accents and punctuation", () => {
    expect(normalizeGigSearchText(" São José / MÚSICA! "))
      .toBe("sao jose musica");
  });

  test("indexes title, description and location with stable grams", () => {
    const grams = buildGigSearchGrams({
      title: "Sax",
      description: "Jazz",
      location: {label: "Niterói"},
    });

    expect(grams).toEqual(expect.arrayContaining([
      "s",
      "sa",
      "sax",
      "j",
      "ja",
      "jaz",
      "nit",
    ]));
    expect(grams).toEqual([...new Set(grams)].sort());
    expect(GIG_SEARCH_SCHEMA_VERSION).toBe(1);
  });

  test("caps indexed text to keep Firestore index entries bounded", () => {
    const uniqueTrigrams = buildGigSearchGrams({
      title: "a".repeat(3000),
      description: "ignored after the cap",
    });

    expect(uniqueTrigrams).toEqual(["a", "aa", "aaa"]);
  });
});
